File: packages/runtimes/src/registry.ts
import type { RuntimeAdapter } from "@dap/types";

export class RuntimeRegistry {
  private adapters = new Map<string, RuntimeAdapter>();

  register(adapter: RuntimeAdapter): void {
    if (this.adapters.has(adapter.id)) {
      throw new Error(`Runtime adapter already registered: ${adapter.id}`);
    }
    this.adapters.set(adapter.id, adapter);
  }

  get(id: string): RuntimeAdapter {
    const adapter = this.adapters.get(id);
    if (!adapter) {
      throw new Error(`Runtime adapter not found: ${id}`);
    }
    return adapter;
  }

  has(id: string): boolean {
    return this.adapters.has(id);
  }

  list(): RuntimeAdapter[] {
    return Array.from(this.adapters.values());
  }
}

export async function createDefaultRegistry(): Promise<RuntimeRegistry> {
  const registry = new RuntimeRegistry();

  const { BashAdapter } = await import("./adapters/bash.js");
  const { HttpAdapter } = await import("./adapters/http.js");
  const { ApiCallAdapter } = await import("./adapters/api-call.js");
  const { ClaudeCodeAdapter } = await import("./adapters/claude-code.js");
  const { GeminiCliAdapter } = await import("./adapters/gemini-cli.js");
  const { CodexAdapter } = await import("./adapters/codex.js");
  const { AiderAdapter } = await import("./adapters/aider.js");

  registry.register(new BashAdapter());
  registry.register(new HttpAdapter());
  registry.register(new ApiCallAdapter());
  registry.register(new ClaudeCodeAdapter());
  registry.register(new GeminiCliAdapter());
  registry.register(new CodexAdapter());
  registry.register(new AiderAdapter());

  return registry;
}

File: packages/runtimes/src/dap_runtimes/registry.py
from __future__ import annotations

from dap_types import RuntimeAdapter

from dap_runtimes.adapters.aider import AiderAdapter
from dap_runtimes.adapters.api_call import ApiCallAdapter
from dap_runtimes.adapters.bash import BashAdapter
from dap_runtimes.adapters.claude_code import ClaudeCodeAdapter
from dap_runtimes.adapters.codex import CodexAdapter
from dap_runtimes.adapters.gemini_cli import GeminiCliAdapter
from dap_runtimes.adapters.http import HttpAdapter


class RuntimeRegistry:
    """Lookup registry dla runtime adapterów."""

    def __init__(self) -> None:
        self._adapters: dict[str, RuntimeAdapter] = {}

    def register(self, adapter: RuntimeAdapter) -> None:
        if adapter.id in self._adapters:
            raise ValueError(f"Runtime adapter already registered: {adapter.id}")
        self._adapters[adapter.id] = adapter

    def get(self, adapter_id: str) -> RuntimeAdapter:
        if adapter_id not in self._adapters:
            raise KeyError(f"Runtime adapter not found: {adapter_id}")
        return self._adapters[adapter_id]

    def has(self, adapter_id: str) -> bool:
        return adapter_id in self._adapters

    def list(self) -> list[RuntimeAdapter]:
        return list(self._adapters.values())


async def create_default_registry() -> RuntimeRegistry:
    """Tworzy registry z 7 wbudowanymi adapterami (F0 = stuby)."""
    registry = RuntimeRegistry()
    registry.register(BashAdapter())
    registry.register(HttpAdapter())
    registry.register(ApiCallAdapter())
    registry.register(ClaudeCodeAdapter())
    registry.register(GeminiCliAdapter())
    registry.register(CodexAdapter())
    registry.register(AiderAdapter())
    return registry
